module/zfs/dsl_dir.c

```diff
@@ -2224,6 +2242,86 @@ dsl_dir_is_zapified(dsl_dir_t *dd)
 	return (doi.doi_type == DMU_OTN_ZAP_METADATA);
 }
 
+void
+dsl_dir_livelist_open(dsl_dir_t *dd, uint64_t obj)
+{
+	objset_t *mos = dd->dd_pool->dp_meta_objset;
+	ASSERT(spa_feature_is_active(dd->dd_pool->dp_spa,
+	    SPA_FEATURE_LIVELIST));
+	dsl_deadlist_open(&dd->dd_livelist, mos, obj);
+	bplist_create(&dd->dd_pending_allocs);
+	bplist_create(&dd->dd_pending_frees);
+}
+
+void
+dsl_dir_livelist_close(dsl_dir_t *dd)
+{
+	dsl_deadlist_close(&dd->dd_livelist);
+	bplist_destroy(&dd->dd_pending_allocs);
+	bplist_destroy(&dd->dd_pending_frees);
+}
+
+void
+dsl_dir_remove_livelist(dsl_dir_t *dd, dmu_tx_t *tx, boolean_t total)
+{
+	uint64_t obj;
+	dsl_pool_t *dp = dmu_tx_pool(tx);
+	spa_t *spa = dp->dp_spa;
+	if (!dsl_deadlist_is_open(&dd->dd_livelist))
+		return;
+	livelist_condense_entry_t to_condense = spa->spa_to_condense;
+
+	/*
+	 * If the livelist being removed is set to be condensed, stop the
+	 * condense zthr and indicate the cancellation in the spa_to_condense
+	 * struct in case the condense no-wait synctask has already started
+	 */
+	zthr_t *ll_condense_thread = spa->spa_livelist_condense_zthr;
+	if (ll_condense_thread != NULL &&
+	    (to_condense.ds != NULL) && (to_condense.ds->ds_dir == dd)) {
+			/*
+			 * We use zthr_finish instead of zthr_cancel because we
+			 * don't want to destroy the zthr, just have it skip
+			 * it's current task.
+			 */
+			spa->spa_to_condense.cancelled = B_TRUE;
+			zthr_finish(ll_condense_thread);
+			/*
+			 * If we've returned from zthr_finish without clearing
+			 * the to_condense data structure it's either because
+			 * the no-wait synctask has started (which is indicated
+			 * by 'syncing' field of to_condense) and we can expect
+			 * it to clear to_condense on its own. Otherwise, we
+			 * returned before the zthr ran. The checkfunc will
+			 * now fail as cancelled == B_TRUE so we can safely NULL
+			 * out ds, allowing a different dir's livelist to be
+			 * condensed.
+			 *
+			 * We can be sure that the to_condense struct will not
+			 * be repopulated during the execution of this function
+			 * because dsl_livelist_try_condense excutes in syncing
+			 * context.
```

Comment:
`... because dsl_livelist_try_condense executes in syncing context.`
may be add at the end of this sentence
`... executes in syncing context, like this function.`
or something along those lines emphasizing your point more that both the current function and `dsl_livelist_try_condense()` are in syncing context, therefore they can't be run concurrently.